HIP/hip-904-frictionless-airdrop.md

```diff
@@ -0,0 +1,711 @@
+
+> ⚠️ `TokenAirdropTransaction` will not support tokens with custom fallback fees that require the receiver to pay the fallback fee.
+
+
+When a sender airdrops a token, if the receiver’s account is already associated with the token, then it is effectively treated as a normal crypto transfer, except that an *airdrop fee* is assessed. This fee exists for all uses of `TokenAirdropTransaction` as a spam-deterrent fee.
+
+If the receiver’s account is **not** associated with the token, but has available auto-association slots, then one of the slots is taken (i.e. `used_auto_associations` is incremented) and the transaction payer will pay for the association fee. They also pay a fee representing the price of a full auto-renewal period’s rent. This is done to make sure the receiver has time to dissociate from the token before renewal takes place. For example, if the airdrop were to happen the minute before account renewal, the token would not included in renewal because it has already been pre-paid by the sender for one full renewal period. Thus, the receiver always has at least one full renewal period to dissociate from any tokens they do not want to own.
+
+If the receiver’s account is **not** associated with the token, and they have **no** free auto-association slots, then the airdrop transfer neither succeeds nor fails, but is held in state as a *pending transfer*.
+
+### Pending Transfers
+
+When a `TokenAirdropTransaction` is handled for a receiver that has no free auto-association slots, a *pending transfer* is created and held in state. Tokens in the pending transfer are not deducted from the sender, nor are they credited to the receiver, until the receiver *claims* the airdrop. If the receiver never claims the airdrop, then the sender continues to pay for it forever, or until their account expires, is deleted, or the sender *cancels* the pending transfer.
+
+To support a users ability to claim a pending transfer a new `TokenClaimAirdrop` transaction will be introduced.
+When a receiver claims a token, the `TokenClaimAirdrop` transaction **must** be signed by the receiver, in addition to the transaction payer (who may be the same as the receiver). In essence, if an account has no free auto-association slots, then it behaves as though receiver-sig-required were true *for airdrops*. That is, the receiver must sign the transaction to receive the airdrop.
+
+A sender may cancel pending transfers for a low nominal fee using the new `TokenCancelAirdrop` transaction. A receiver can also choose to just ignore pending transfers. It is expected that wallets may build special user workflows into their apps to allow customers to ignore pending transfers without having to actually issue a transaction to the network to cancel them. Since pending transfers are paid for by the sender, the receiver may choose not to simply filter them out in their wallet software.
+
+All pending transfers sent by an account must be canceled before the account can be deleted.
```

Comment:
I read "sent" as either the payer for the airdrop transaction, or the "treasury" account holding the air dropped tokens, should we clarify which here ?

Reply:
Below, we state that if an account fails to renew rent, pending airdrops are automatically deleted, why have a different behaviour on account deletion ? Would it not be easier to just cancel all pending airdrops on deletion and not impose this requirement on account holders ?

Reply:
(I'll make this comment here rather than below, because it fits with Greg's comment:). If _all_ pending airdrops from an account are deleted when the account fails to renew rent is that a potential latency issue if there are a _lot_ of them accumulated?  They were only added 10 at a time (due to the the transfer limit there) but you might have to delete thousands in the time of one transaction ...

(Although I do see a possibility: Deleting state when rent is unpaid is at _our_ discretion and we could have a mechanism to spread out the work as we see fit.  Though if we did that they'd still have to be ignored, though part of the state.  But that has a fixed cost per pending tx at the time of a query or claim by a receiving account.  That same lazy mechanism, with the ability to ignore pending transactions based on whether the account is valid, would work for delete as well.)

Reply:
Another comment:  If, to delete an account you have to delete all pending airdrops, and that takes transaction fees, could this not lead to _abandoned_ accounts?  User just says "heck with it, see ya!", and the stuff is in our state until we implement rent and time it out.  (Or are there other reasons for the user to delete his account (and clean things up) so he doesn't do this?)